sphinx/user-guide/configuration.rst

```diff
@@ -0,0 +1,873 @@
+
+      opts=ketchup mayonnaise
+
+implies loading the optional configuration ``rose-app-ketchup.conf`` and then
+the optional configuration ``rose-app-mayonnaise.conf``, which may override
+the previous one.
+
+By default, a Rose command will fail if an optional configuration file is
+missing. However, if you put the optional configuration key in brackets,
+then the optional configuration file is allowed to be missing. E.g.:
+
+   .. code-block:: rose
+
+      opts=ketchup (mayonnaise)
+
+In the above example, ``rose-app-mayonnaise.conf`` can be missing.
+
+Some Rose utilities (e.g. :ref:`command-rose-suite-run`,
+:ref:`command-rose-task-run`, :ref:`command-rose-app-run`, etc) allow
+optional configurations to be selected at run time using:
+
+#. The ``ROSE_APP_OPT_CONF_KEYS`` Environment variables.
+#. The command line options ``--opt-conf-key=KEY`` or ``-O KEY``.
+
+.. tip::
+   See reference of individual commands for detail.
+
+.. note::
+   By default optional configurations must exist else an error will
+   be raised. To specify an optional configuration which may be missing write
+   the name of the configuration inside parenthesis (e.g. ``(foo)``).
+
+Optional Configurations and Metadata
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Metadata utilities such as ``rose app-upgrade`` and ``rose macro`` treat each
+main + optional configuration as a separate entity to be transformed,
+upgraded, or validated. Use cases with more than one optional configuration
+are not handled.
+
+When transforming or upgrading, each optional configuration is treated
+separately and re-created after the transform as a functional difference
+from the main upgraded configuration.
+
+The logic for transforming or upgrading a main configuration ``C`` with
+optional configurations ``O1`` and ``O2`` into a new main configuration ``Ct``
+and new optional configurations ``O1t`` and ``O2t`` can be represented like
+this:
+
+   .. code-block:: none
+
+      C => Ct
+      C + O1 => C1t
+      C + O2 => C2t
+      O1t = C1t - Ct
+      O2t = C2t - Ct
+
+
+Import Configuration
+--------------------
+
+A root level ``import=PATH1 PATH2...`` setting in the main configuration will
+tell Rose utilities to search for configurations at ``PATH1``, ``PATH2`` (and
+so on) and inherit configuration and files from them if found.
+
+.. tip::
+   At the moment, use of this is only encouraged for configuration metadata.
+
+
+Re-define Configuration at Run Time
+-----------------------------------
+
+Some Rose utilities (e.g. :ref:`command-rose-suite-run`,
+:ref:`command-rose-task-run`, :ref:`command-rose-app-run`, etc) allow you
+to re-define configuration settings at run time using the
+``--define=[SECTION]NAME=VALUE`` or ``-D [SECTION]NAME=VALUE`` options on
+the command line. This would add new settings or override any settings
+defined in the main and optional configurations. E.g.:
+
+   .. code-block:: bash
+
+      (shell)$ # Set [env]FOO=foo, and [env]BAR=bar
```

Comment:
No problem, I will remove those - I simply left them in because I was trying to make as few changes other than restructuring to the raw copied text as possible, but it would look better and be more consistent without these, certainly.